src/IrcConnection/IrcConnection.cs

```diff
@@ -940,38 +687,29 @@ public void WriteLine(string data)
             WriteLine(data, Priority.Medium);
         }
 
-        private bool _WriteLine(string data)
+        /// <summary>
+        /// Received from the write thread when a line is successfully written to the connection
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void _OnWriteLine(object sender, WriteLineEventArgs e)
         {
-            if (IsConnected) {
-                try {
-                    lock (_Writer) {
-                        _Writer.Write(data + "\r\n");
-                        _Writer.Flush();
-                    }
-                } catch (IOException) {
-#if LOG4NET
-                    Logger.Socket.Warn("sending data failed, connection lost");
-#endif
-                    IsConnectionError = true;
-                    return false;
-                } catch (ObjectDisposedException) {
-#if LOG4NET
-                    Logger.Socket.Warn("sending data failed (stream error), connection lost");
-#endif
-                    IsConnectionError = true;
-                    return false;
-                }
-
-#if LOG4NET
-                Logger.Socket.Debug("sent: \""+data+"\"");
-#endif
-                if (OnWriteLine != null) {
-                    OnWriteLine(this, new WriteLineEventArgs(data));
-                }
-                return true;
+            if (OnWriteLine != null) {
+                OnWriteLine(sender, e);
             }
+        }
 
-            return false;
+        /// <summary>
+        /// Received from the transport when a line is successfully received from the connection
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void Transport_OnMessageReceived(object sender, ReadLineEventArgs e)
+        {
+            // Add it to the read queue
+            // Do we need lock (this) here?
```

Comment:
This depends on the thread that raises the event and if _ReadQueue is thread-safe or not